net/mwan3/files/usr/sbin/mwan3

```diff
@@ -37,52 +38,64 @@ ifdown()
 
 	ACTION=ifdown INTERFACE=$1 /sbin/hotplug-call iface
 
-	kill $(pgrep -f "mwan3track $1 $2") &> /dev/null
+	kill $(pgrep -f "mwan3track $1 ") &> /dev/null
 	mwan3_track_clean $1
 }
 
 ifup()
 {
-	local device enabled up l3_device status
-
-	mwan3_lock "command" "mwan3"
-
-	config_load mwan3
-	config_get_bool enabled globals 'enabled' 0
-
-	[ ${enabled} -gt 0 ] || {
-		echo "The service mwan3 is global disabled."
-		echo "Please execute \"/etc/init.d/mwan3 start\" first."
-		mwan3_unlock "command" "mwan3"
-		exit 1
-	}
+	local device enabled up l3_device status interface true_iface
 
 	if [ -z "$1" ]; then
 		echo "Expecting interface. Usage: mwan3 ifup <interface>"
-		mwan3_unlock "command" "mwan3"
 		exit 0
 	fi
 
 	if [ -n "$2" ]; then
 		echo "Too many arguments. Usage: mwan3 ifup <interface>"
-		mwan3_unlock "command" "mwan3"
 		exit 0
 	fi
 
-	config_get enabled "$1" enabled 0
-	mwan3_unlock "command" "mwan3"
+	interface=$1
+
+	if [ "${MWAN3_STARTUP}" != 1 ]; then
+		# It is not necessary to obtain a lock here, because it is obtained in the hotplug
+		# script, but we still want to do the check to print a useful error message
+		config_load mwan3
+		config_get_bool enabled globals 'enabled' 0
+
+		[ ${enabled} -gt 0 ] || {
+			echo "The service mwan3 is global disabled."
+			echo "Please execute \"/etc/init.d/mwan3 start\" first."
+			exit 1
+		}
+	else
+		enabled=1
+	fi
+	mwan3_get_true_iface true_iface $interface
+	status=$(ubus -S call network.interface.$true_iface status)
 
-	status=$(ubus -S call network.interface.$1 status)
 	[ -n "$status" ] && {
 		json_load "$status"
 		json_get_vars up l3_device
 	}
+	hotplug_startup()
+	{
+		MWAN3_STARTUP=$MWAN3_STARTUP ACTION=ifup INTERFACE=$true_iface DEVICE=$l3_device sh /etc/hotplug.d/iface/15-mwan3
+		MWAN3_STARTUP=$MWAN3_STARTUP ACTION=ifup INTERFACE=$true_iface DEVICE=$l3_device sh /etc/hotplug.d/iface/16-mwan3-user
```

Comment:
here
`INTERFACE=$true_iface`
should be "$interface" ?

Reply:
Yes. I think you are correct here. The current implementation is actually a bit broken. If the rule is defined on e.g. 'WAN' but the actual interface is named 'WAN_6' then when 'WAN_6' has a hotplug event the mwan3 hotplug script will not run. If WAN has a hotplug event and WAN_6 does not, the hotplug event will run unnecessarily.

It actually can be even worse than that, if a user has WAN and WAN_4 that are different interfaces (not sure why, maybe they have WAN, WAN_2, WAN_3 and WAN_4?), then mwan3 will return the ip address for WAN_4 for the WAN interface. To be clear, this is already broken in the current version, and not a new issue with this P/R. I made an issue for it, but I think fixing that is beyond the scope of this P/R. I made an issue for it here #13133 

Best practice is to have the mwan3 interface name match the openwrt interface name. Still, it is bad user experience to break things, even if the features are undocumented. This P/R is focusing on performance improvements - having the extra `ubus` calls has a very minor performance hit, but it is probably worth the performance penalty to not break existing configurations.

Priorities for P/Rs should be:
1) Not introducing new bugs for people running things set up following the documentation
2) Fixing existing bugs
3) Adding new features and improvements
4) Not breaking undocumented behavior that people could be using

Reply:
ok for now.